### ISSUE TITLE
wait_dequeue_timed don't return.

### ISSUE THREAD

**Issue by horllion** [cameron314/concurrentqueue#173]
produce with Multiple threads, using moodycamel::BlockingConcurrentQueue::enqueue。
only one single thread consume with using   wait_dequeue_timed, timeOut param value is 3000 microseconds.
stack is in while ( !inner.try_dequeue( item ) ).
![image](https://user-images.githubusercontent.com/42425728/69220106-32578400-0baf-11ea-9509-9d58b0c1f202.png)
I doubt if the size_approx return 0, try_dequeue return false, wait_dequeue_timed is in while loop.

**Comment by horllion** [comment]
![image](https://user-images.githubusercontent.com/42425728/69220463-038ddd80-0bb0-11ea-83a2-d4ac46e4c0ab.png)
should is this tailIndex.load( std::memory_order_acquire);

**Comment by horllion** [comment]
I think I found this bug。
![image](https://user-images.githubusercontent.com/42425728/69225850-1c02f580-0bba-11ea-88b2-9a2bf2751cee.png)
2 or more threads get the same currentTailIndex，one of them lost, but queue return true.
![image](https://user-images.githubusercontent.com/42425728/69226066-81ef7d00-0bba-11ea-8c10-777435e10134.png)
so sema singal, as m_count > 0 but size_approx() return 0;

**Comment by cameron314** [comment]
Internal sub-queues are SPMC -- two threads are never manipulating the `tailIndex` of the same object at once. Though I've suspected for a while there might be a bug in the semaphore code.

Do you have a simple example that can reproduce the problem?

**Comment by horllion** [comment]
> Internal sub-queues are SPMC -- two threads are never manipulating the `tailIndex` of the same object at once. Though I've suspected for a while there might be a bug in the semaphore code.
> 
> Do you have a simple example that can reproduce the problem?

my blocked Queue code:

class BlockedQueue
{
public:
    BlockedQueue( int capacity );
    virtual ~BlockedQueue();

public:
    bool put( Poco::SharedPtr< RecogCmd > data );

    Poco::SharedPtr< RecogCmd > take( int timeOut );

    size_t size();
    bool   isFull();

private:
    Poco::Mutex                                                        _mutex;
    Poco::Condition                                                    _notEmpty;
    moodycamel::BlockingConcurrentQueue< Poco::SharedPtr< RecogCmd > > _queue;
    int                                                                _capacity;
};
55 threads randomly put:
bool BlockedQueue::put( Poco::SharedPtr< RecogCmd > data )
{
    // Poco::Mutex::ScopedLock lock(_mutex);
    return _queue.enqueue( data );
}
one thread loop take:

Poco::SharedPtr< RecogCmd > BlockedQueue::take( int timeOut )
{
    Poco::SharedPtr< RecogCmd > frame;
    _queue.wait_dequeue_timed( frame, timeOut * 1000 );  // timeOut 毫秒, wait_dequeue_timed 参数微妙
                                                         //_queue.wait_dequeue(frame);
    return frame;
}
while (m_isRun)
    {
        bool                        bTimeout = false;
        Poco::SharedPtr< RecogCmd > item = m_blockedQueue->take(averageTimeout);
        if ( item.isNull() )
        {
           ……
            bTimeout = true;
        }
        else
        {
            vCmd.push_back(item);
        }

        int size = vCmd.size();
        if (  bTimeout || ( size % nMaxBatchSize == 0 ) )
        {
            ……

            std::unique_lock< std::mutex > theLocker(mutex);
            if(size > 0)
            {
               ……
            }
            theLocker.unlock();
            cond.notify_one();
        }
    }
consumer log print bTimeout is always false, and pstack prove theLocker is not locked.

**Comment by horllion** [comment]
It seems that the shorter the timeout, the easier it is to appear.

**Comment by horllion** [comment]
moodycamel::BlockingConcurrentQueue< std::string > _queue;
    std::vector<std::thread> vecs;
    std::atomic<uint64_t> count{0};
    for(int i = 0; i < 100; i++){
        std::thread recogThr([&](){
            while (true)
            {
                std::string str;
                int64_t begin = time(NULL);
                _queue.wait_dequeue_timed( str, 5000 );
                int64_t end = time(NULL);
                if(end - begin >= 2)
                {
                    printf("time eslaped more than one secs, after count: %lu\n", count.load());
                }
                count++;
            }
        });
        recogThr.detach();
    }
    while (true)
    {
        std::string str;
        int64_t begin = time(NULL);
        _queue.wait_dequeue_timed( str, 5000 );
        int64_t end = time(NULL);
        if(end - begin >= 2)
        {
            printf("time eslaped more than one secs, after count: %lu\n", count.load());
        }
        count++;
    }

**Comment by horllion** [comment]
I modified timed_wait
![image](https://user-images.githubusercontent.com/42425728/70029830-e65f0300-15e2-11ea-98f4-da35439fa952.png)
reproduced this bug，I get the errno is ENINVAL,but timed_wait return true.
![image](https://user-images.githubusercontent.com/42425728/70029920-1908fb80-15e3-11ea-8169-192fc8f59c66.png)
Maybe clock_gettime failed caused.

**Comment by cameron314** [comment]
Thanks for taking the time to track this down, this indeed looks like a bug in the semaphore logic where it determines whether the call succeeds or not.

I'm currently away from my PC but I'll try to find time to fix this next week.

Sorry for the long response time on my end.
